2021/ruby/day4/lib/day4.rb

```diff
@@ -0,0 +1,109 @@
+class BingoGame
+  attr_accessor :drawn_numbers, :boards
+
+  def initialize(drawn_numbers, boards)
+    @drawn_numbers = drawn_numbers
+    @boards = boards
+  end
+
+  def self.from_file(filepath)
+    raw_content = File.read(filepath, chomp: true).split("\n\n")
+
+    drawn_numbers = raw_content.shift.split(",").map(&:to_i)
+    boards = raw_content.map { |raw_board| BingoBoard.from_raw(raw_board) }
+
+    new(drawn_numbers, boards)
+  end
+
+  def part1
+    @drawn_numbers.each do |number|
+      @boards.each do |board|
+        board.call_number(number)
+        if board.has_win?
+          return board.final_score(number)
+        end
+      end
+    end
+  end
+
+  def part2
+    boards_in_play = @boards.clone
+    @drawn_numbers.each do |number|
+      @boards.each do |board|
+        board.call_number(number)
+
+        if !board.won && board.has_win?
+          if boards_in_play.one?
+            return board.final_score(number)
+          end
+          board.won = true
+          boards_in_play.delete(board)
+        end
+      end
+    end
+  end
+end
+
+class BingoBoard
+  attr_accessor :board, :won
+
+  def initialize(board)
+    @board = board
+    @won = false
+  end
+
+  def self.from_raw(raw_board)
+    new(
+      raw_board.split(/\n/).map do |line|
+        line.split.map { |raw_number| {value: raw_number.to_i, called: false} }
+      end
+    )
+  end
+
+  def call_number(number)
+    @board.each do |row|
+      row.each do |spot|
+        spot[:called] = true if number == spot[:value]
+      end
+    end
+  end
+
+  def uncalled_numbers
+    uncalled = []
+    @board.each do |row|
+      row.each do |spot|
+        uncalled.push(spot[:value]) if spot[:called] == false
+      end
+    end
+    uncalled
+  end
```

Comment:
This is the one place where I couldn't get Ruby's great `map` or `reduce` methods to work. `map` **seems like** it should work here, but I tried several combinations that didn't work. It's a little tricker with two loops. Didn't seem worth the time to keep digging into.